snappy/snappy.go

```diff
@@ -0,0 +1,88 @@
+package snappy
+
+import (
+	"bytes"
+	"encoding/binary"
+
+	"github.com/golang/snappy"
+	kafka "github.com/segmentio/kafka-go"
+)
+
+func init() {
+	kafka.RegisterCompressionCodec(func() kafka.CompressionCodec {
+		return NewCompressionCodec()
+	})
+}
+
+type CompressionCodec struct {
+	writer *snappy.Writer
+}
+
+const Code = 2
+
+func NewCompressionCodec() CompressionCodec {
+	return CompressionCodec{}
+}
+
+// Code implements the kafka.CompressionCodec interface.
+func (c CompressionCodec) Code() int8 {
+	return Code
+}
+
+func (c CompressionCodec) Encode(dst, src []byte) (int, error) {
+	buf := buffer{data: dst}
+	return buf.Write(snappy.Encode(nil, src))
```

Comment:
Instead of using `buffer` and passing `nil` to `snappy.Encode` wouldn't it make more sense to just use `snappy.MaxEncodedLen` to check if the data fits? This way `dst` won't be changed if it's too small and if it fits you can just pass `dst` to `snappy.Encode`, avoiding an allocation.

E.g. 

```go
if n := snappy.MaxEncodedLen(len(src)); n > len(dst) {
    return 0, bytes.ErrTooLarge
}

return snappy.Encode(dst, src)
```

Reply:
That's a good suggestion. I think we're gonna try to focus on an OK implementation in this PR, then we can think about ways to make it faster. I'll leave it up to @Pryz to make the call here tho since he's leading the effort.